lib/pbench/test/functional/agent/cli/commands/tools/test_tools_list.py

```diff
@@ -1,31 +1,106 @@
 import pytest
 
 
-def test_pbench_list_tools_help():
-    command = ["pbench-list-tools", "--help"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert b"Usage: pbench-list-tools [OPTIONS]" in out
-    assert exitcode == 0
-
-
-def test_list_tool(monkeypatch, agent_config, pbench_run, pbench_cfg):
-    monkeypatch.setenv("_PBENCH_AGENT_CONFIG", str(pbench_cfg))
-    p = pbench_run / "tools-v1-default" / "testhost.example.com"
-    p.mkdir(parents=True)
-    tool = p / "perf"
-    tool.touch()
-
-    command = ["pbench-list-tools"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "--group", "default"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"default: testhost.example.com ['perf']" in out
-
-    command = ["pbench-list-tools", "-n", "perf"]
-    out, err, exitcode = pytest.helpers.capture(command)
-    assert exitcode == 0
-    assert b"tool name: perf groups: default" in out
+EMPTY = b""
+USAGE = b"Usage: pbench-list-tools [OPTIONS]"
+BAD_GROUP = b"Bad tool group: "
+TRACEBACK = b"Traceback (most recent call last):\n"
+
+
+class Test_list_tools_no_tools_registered:
+    def test_help(self):
+        command = ["pbench-list-tools", "--help"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert TRACEBACK not in err
+        assert EMPTY == err
+        assert USAGE in out
+        assert exitcode == 0
+
+    def test_no_args(self, agent_config):
+        command = ["pbench-list-tools"]
+        out, err, exitcode = pytest.helpers.capture(command)
+        assert TRACEBACK not in err
+        assert EMPTY == err
+        assert EMPTY == out
+        assert exitcode == 0
+
+
+class Test_list_tools_tools_registered:
+    @pytest.fixture
+    def tool(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        tool = p / "perf"
+        tool.touch()
+
+    @pytest.fixture
+    def tool_with_random_file(self, pbench_run):
+        p = pbench_run / "tools-v1-default" / "testhost.example.com"
+        p.mkdir(parents=True)
+        (p.parent / "foo").touch()
```

Comment:
Should we be testing "random files" at other directory levels, too? E.g., trigger/label file skipping?

Reply:
Yes, I think so: although most of the test cases arose either out of know bugs or of testing "normal" functionality, there are still a lot of corner cases to test. Actually, I think that the current implementation of `pbench-list-tools` is going to be replaced: the only thing that will be left will be the test suite, so the more complete that is, the better off we'll be. I imagine the test suite  will grow and, in some sense, be a model of what the test suites for other tools will look like.